File: app/src/main/java/ru/p3tr0vich/fuel/FragmentPreferences.java
package ru.p3tr0vich.fuel;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.support.v7.preference.EditTextPreference;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceGroup;
import android.support.v7.preference.PreferenceScreen;
import android.text.TextUtils;

import ru.p3tr0vich.fuel.utils.UtilsFormat;
import ru.p3tr0vich.fuel.utils.UtilsLog;

import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.DEFAULT_COST;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.DEFAULT_VOLUME;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.MAP_CENTER_TEXT;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.PRICE;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.SMS;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.SMS_ADDRESS;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.SMS_ENABLED;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.SMS_TEXT_PATTERN;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.SYNC;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.SYNC_ENABLED;
import static ru.p3tr0vich.fuel.helpers.PreferencesHelper.Keys.SYNC_YANDEX_DISK;

public class FragmentPreferences extends FragmentPreferencesBase implements
        SharedPreferences.OnSharedPreferenceChangeListener,
        Preference.OnPreferenceClickListener {

    public static final String TAG = "FragmentPreferences";

    private static final boolean LOG_ENABLED = false;

    public static final String KEY_PREFERENCE_SCREEN = "KEY_PREFERENCE_SCREEN";

    private OnPreferenceScreenChangeListener mOnPreferenceScreenChangeListener;
    private OnPreferenceSyncEnabledChangeListener mOnPreferenceSyncEnabledChangeListener;
    private OnPreferenceClickListener mOnPreferenceClickListener;

    private boolean mIsInRoot;
    private PreferenceScreen mRootPreferenceScreen;

    public interface OnPreferenceClickListener {
        void onPreferenceMapCenterClick();

        void onPreferenceSyncYandexDiskClick();

        void onPreferenceSMSAddressClick();

        void onPreferenceSMSTextPatternClick();
    }

    public interface OnPreferenceScreenChangeListener {
        void onPreferenceScreenChanged(@NonNull CharSequence title, boolean isInRoot);
    }

    public interface OnPreferenceSyncEnabledChangeListener {
        void onPreferenceSyncEnabledChanged(boolean enabled);
    }

    @NonNull
    public static Fragment newInstance(int id) {
        return newInstance(id, new FragmentPreferences());
    }

    @NonNull
    @Override
    public String getTitle() {
        return mIsInRoot ? getString(R.string.title_prefs) : (String) getPreferenceScreen().getTitle();
    }

    public boolean isInRoot() {
        return mIsInRoot;
    }

    @SuppressLint("SwitchIntDef")
    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (LOG_ENABLED) UtilsLog.d(TAG, "onSharedPreferenceChanged", "key == " + key);

        updatePreference(key);

        switch (preferencesHelper.keys.getAsInt(key)) {
            case PRICE:
                updatePreference(preferencesHelper.keys.defaultCost);
                updatePreference(preferencesHelper.keys.defaultVolume);
                break;
            case DEFAULT_COST:
                updatePreference(preferencesHelper.keys.defaultVolume);
                break;
            case DEFAULT_VOLUME:
                updatePreference(preferencesHelper.keys.defaultCost);
                break;
            case SYNC_ENABLED:
                updatePreference(preferencesHelper.keys.sync);
                mOnPreferenceSyncEnabledChangeListener.onPreferenceSyncEnabledChanged(
                        preferencesHelper.isSyncEnabled());
                break;
            case SMS_ENABLED:
                updatePreference(preferencesHelper.keys.sms);
        }
    }

    @SuppressLint("SwitchIntDef")
    @Override
    public boolean onPreferenceClick(Preference preference) {
        switch (preferencesHelper.keys.getAsInt(preference.getKey())) {
            case MAP_CENTER_TEXT:
                mOnPreferenceClickListener.onPreferenceMapCenterClick();
                return true;
            case SYNC_YANDEX_DISK:
                mOnPreferenceClickListener.onPreferenceSyncYandexDiskClick();
                return true;
            case SMS_ADDRESS:
                mOnPreferenceClickListener.onPreferenceSMSAddressClick();
                return true;
            case SMS_TEXT_PATTERN:
                mOnPreferenceClickListener.onPreferenceSMSTextPatternClick();
                return true;
            default:
                return false;
        }
    }

    @Override
    public void onCreatePreferences(Bundle bundle, String s) {
        if (LOG_ENABLED) UtilsLog.d(TAG, "onCreatePreferences");

        addPreferencesFromResource(R.xml.preferences);
        mRootPreferenceScreen = getPreferenceScreen();

        findPreference(preferencesHelper.keys.mapCenterText).setOnPreferenceClickListener(this);
        findPreference(preferencesHelper.keys.syncYandexDisk).setOnPreferenceClickListener(this);
        findPreference(preferencesHelper.keys.smsAddress).setOnPreferenceClickListener(this);
        findPreference(preferencesHelper.keys.smsTextPattern).setOnPreferenceClickListener(this);

        String keyPreferenceScreen = null;

        if (bundle == null) {
            Bundle arguments = getArguments();
            if (arguments != null)
                keyPreferenceScreen = arguments.getString(KEY_PREFERENCE_SCREEN);
        } else
            keyPreferenceScreen = bundle.getString(KEY_PREFERENCE_SCREEN);

        if (TextUtils.isEmpty(keyPreferenceScreen))
            mIsInRoot = true;
        else
            navigateToScreen((PreferenceScreen) findPreference(keyPreferenceScreen));
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putString(KEY_PREFERENCE_SCREEN, mIsInRoot ? null : getPreferenceScreen().getKey());
    }

    @Override
    public void onNavigateToScreen(PreferenceScreen preferenceScreen) {
        navigateToScreen(preferenceScreen);

        super.onNavigateToScreen(preferenceScreen);
    }

    private void navigateToScreen(@Nullable PreferenceScreen preferenceScreen) {
        mIsInRoot = preferenceScreen == null || preferenceScreen.equals(mRootPreferenceScreen);

        setPreferenceScreen(mIsInRoot ? mRootPreferenceScreen : preferenceScreen);

        mOnPreferenceScreenChangeListener.onPreferenceScreenChanged(getTitle(), mIsInRoot);
    }

    public boolean goToRootScreen() {
        if (mIsInRoot) return false;

        navigateToScreen(null);

        return true;
    }

    public void goToSyncScreen() {
        navigateToScreen((PreferenceScreen) findPreference(preferencesHelper.keys.sync));
    }

    @Override
    public boolean onBackPressed() {
        return goToRootScreen();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mOnPreferenceScreenChangeListener = (OnPreferenceScreenChangeListener) context;
            mOnPreferenceSyncEnabledChangeListener = (OnPreferenceSyncEnabledChangeListener) context;
            mOnPreferenceClickListener = (OnPreferenceClickListener) context;
        } catch (ClassCastException e) {
            throw new ImplementException(context, new Class[]{
                    OnPreferenceScreenChangeListener.class,
                    OnPreferenceSyncEnabledChangeListener.class,
                    OnPreferenceClickListener.class});
        }
    }

    @Override
    public void onStart() {
        if (LOG_ENABLED) UtilsLog.d(TAG, "onStart");

        super.onStart();

        init(mRootPreferenceScreen);

        mRootPreferenceScreen.getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onStop() {
        if (LOG_ENABLED) UtilsLog.d(TAG, "onStop");

        mRootPreferenceScreen.getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);

        super.onStop();
    }

    private String getValue(@Nullable String value, @NonNull String empty) {
        return TextUtils.isEmpty(value) ? empty : value;
    }

    private String getValue(@Nullable String value, @StringRes int emptyId) {
        return getValue(value, getString(emptyId));
    }

    private void updatePreference(String key) {
        updatePreference(mRootPreferenceScreen.findPreference(key));
    }

    @SuppressLint("SwitchIntDef")
    private void updatePreference(@Nullable Preference preference) {
        if (LOG_ENABLED) UtilsLog.d(TAG, "updatePreference", "preference == " + preference);

        if (preference == null) return;

        String key = preference.getKey();
        if (key == null) return;

        String title = null;
        String summary = null;

        int intKey = preferencesHelper.keys.getAsInt(key);

        switch (intKey) {
            case PRICE:
                summary = getValue(preferencesHelper.getPriceAsString(), R.string.pref_price_empty);
                break;
            case DEFAULT_COST:
            case DEFAULT_VOLUME:
                float cost = preferencesHelper.getDefaultCost();
                float volume = preferencesHelper.getDefaultVolume();

                if (cost == 0 && volume == 0)
                    summary = intKey == DEFAULT_COST ?
                            getString(R.string.pref_def_cost_empty) :
                            getString(R.string.pref_def_volume_empty);
                else {
                    float price = preferencesHelper.getPrice();

                    if (intKey == DEFAULT_COST) {
                        if (cost == 0) {
                            summary = getString(R.string.pref_def_cost_calc);

                            if (price != 0)
                                summary += " (" +  UtilsFormat.floatToString(volume * price) + ")";

                        }
                    } else {
                        if (volume == 0) {
                            summary = getString(R.string.pref_def_volume_calc);

                            if (price != 0)
                                summary += " (" + UtilsFormat.floatToString(cost / price) + ")";
                        }
                    }
                }
                if (TextUtils.isEmpty(summary)) summary = preferencesHelper.getString(key);
                break;
            case MAP_CENTER_TEXT:
                summary = preferencesHelper.getMapCenterText();
                break;
            case SYNC:
                summary = getString(preferencesHelper.isSyncEnabled() ?
                        R.string.pref_sync_summary_on : R.string.pref_sync_summary_off);
                break;
            case SYNC_ENABLED:
                title = getString(preferencesHelper.isSyncEnabled() ?
                        R.string.pref_sync_summary_on : R.string.pref_sync_summary_off);
                break;
            case SMS:
                summary = getString(preferencesHelper.isSMSEnabled() ?
                        R.string.pref_sms_summary_on : R.string.pref_sms_summary_off);
                break;
            case SMS_ENABLED:
                title = getString(preferencesHelper.isSMSEnabled() ?
                        R.string.pref_sms_summary_on : R.string.pref_sms_summary_off);
                break;
            case SMS_ADDRESS:
                summary = preferencesHelper.getSMSAddress();
                break;
            case SMS_TEXT_PATTERN:
                summary = preferencesHelper.getSMSTextPattern();
                summary = summary.replaceAll("[\\s]+", " ");
                break;
            default:
                if (preference instanceof EditTextPreference) {
                    String text = preferencesHelper.getString(key);

                    ((EditTextPreference) preference).setText(text);

                    text = getValue(text, "0");

                    summary = preference.getSummary().toString();

                    int i = summary.lastIndexOf(" (");
                    if (i != -1) summary = summary.substring(0, i);
                    summary += " (" + text + ")";
                }
        }

        if (LOG_ENABLED)
            UtilsLog.d(TAG, "updatePreference", "title == " + title + ", summary == " + summary);

        if (title != null) preference.setTitle(title);
        if (summary != null) preference.setSummary(summary);
    }

    private void init(@NonNull Preference preference) {
        updatePreference(preference);
        if (preference instanceof PreferenceScreen || preference instanceof PreferenceGroup) {
            for (int i = 0, count = ((PreferenceGroup) preference).getPreferenceCount(); i < count; i++)
                init(((PreferenceGroup) preference).getPreference(i));
        }
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/fragments/FragmentPreferences.kt
package ru.p3tr0vich.fuel.fragments

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.text.TextUtils
import androidx.preference.EditTextPreference
import androidx.preference.Preference
import androidx.preference.PreferenceGroup
import androidx.preference.PreferenceScreen
import ru.p3tr0vich.fuel.ImplementException
import ru.p3tr0vich.fuel.R
import ru.p3tr0vich.fuel.factories.FragmentFactory
import ru.p3tr0vich.fuel.utils.UtilsFormat
import ru.p3tr0vich.fuel.utils.UtilsLog

class FragmentPreferences : FragmentPreferencesBase(FragmentFactory.Ids.PREFERENCES),
        SharedPreferences.OnSharedPreferenceChangeListener, Preference.OnPreferenceClickListener {

    private var onPreferenceScreenChangeListener: OnPreferenceScreenChangeListener? = null
    private var onPreferenceSyncEnabledChangeListener: OnPreferenceSyncEnabledChangeListener? = null
    private var onPreferenceClickListener: OnPreferenceClickListener? = null

    var isInRoot: Boolean = false
        private set

    private lateinit var rootPreferenceScreen: PreferenceScreen

    override val title: String
        get() = if (isInRoot) getString(R.string.title_prefs) else preferenceScreen.title as String

    interface OnPreferenceClickListener {
        fun onPreferenceMapCenterClick()

        fun onPreferenceSyncYandexDiskClick()

        fun onPreferenceSMSAddressClick()

        fun onPreferenceSMSTextPatternClick()
    }

    interface OnPreferenceScreenChangeListener {
        fun onPreferenceScreenChanged(title: CharSequence, isInRoot: Boolean)
    }

    interface OnPreferenceSyncEnabledChangeListener {
        fun onPreferenceSyncEnabledChanged(enabled: Boolean)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        if (LOG_ENABLED) {
            UtilsLog.d(TAG, "onSharedPreferenceChanged", "key == $key")
        }

        rootPreferenceScreen = preferenceScreen

        updatePreference(key)

        when (key) {
            preferencesHelper.keys.price -> {
                updatePreference(preferencesHelper.keys.defaultCost)
                updatePreference(preferencesHelper.keys.defaultVolume)
            }
            preferencesHelper.keys.defaultCost -> updatePreference(preferencesHelper.keys.defaultVolume)
            preferencesHelper.keys.defaultVolume -> updatePreference(preferencesHelper.keys.defaultCost)
            preferencesHelper.keys.syncEnabled -> {
                updatePreference(preferencesHelper.keys.sync)

                onPreferenceSyncEnabledChangeListener?.onPreferenceSyncEnabledChanged(
                        preferencesHelper.isSyncEnabled)
            }
            preferencesHelper.keys.smsEnabled -> updatePreference(preferencesHelper.keys.sms)
        }
    }

    override fun onPreferenceClick(preference: Preference): Boolean {
        when (preference.key) {
            preferencesHelper.keys.mapCenterText -> {
                onPreferenceClickListener?.onPreferenceMapCenterClick()
                return true
            }
            preferencesHelper.keys.syncYandexDisk -> {
                onPreferenceClickListener?.onPreferenceSyncYandexDiskClick()
                return true
            }
            preferencesHelper.keys.smsAddress -> {
                onPreferenceClickListener?.onPreferenceSMSAddressClick()
                return true
            }
            preferencesHelper.keys.smsTextPattern -> {
                onPreferenceClickListener?.onPreferenceSMSTextPatternClick()
                return true
            }
            else -> return false
        }
    }

    override fun onCreatePreferences(bundle: Bundle?, s: String?) {
        if (LOG_ENABLED) {
            UtilsLog.d(TAG, "onCreatePreferences", "preferenceScreen == $preferenceScreen")
        }

        addPreferencesFromResource(R.xml.preferences)

        rootPreferenceScreen = preferenceScreen

        findPreference(preferencesHelper.keys.mapCenterText).onPreferenceClickListener = this
        findPreference(preferencesHelper.keys.syncYandexDisk).onPreferenceClickListener = this
        findPreference(preferencesHelper.keys.smsAddress).onPreferenceClickListener = this
        findPreference(preferencesHelper.keys.smsTextPattern).onPreferenceClickListener = this

        val keyPreferenceScreen: String? =
                bundle?.getString(KEY_PREFERENCE_SCREEN)
                        ?: arguments?.getString(KEY_PREFERENCE_SCREEN)

        if (TextUtils.isEmpty(keyPreferenceScreen)) {
            isInRoot = true
        } else {
            navigateToScreen(findPreference(keyPreferenceScreen) as PreferenceScreen)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putString(KEY_PREFERENCE_SCREEN, if (isInRoot) null else preferenceScreen.key)
    }

    override fun onNavigateToScreen(preferenceScreen: PreferenceScreen) {
        navigateToScreen(preferenceScreen)

        super.onNavigateToScreen(preferenceScreen)
    }

    private fun navigateToScreen(preferenceScreen: PreferenceScreen?) {
        isInRoot = preferenceScreen == null || preferenceScreen == rootPreferenceScreen

        setPreferenceScreen(if (isInRoot) rootPreferenceScreen else preferenceScreen)

        onPreferenceScreenChangeListener?.onPreferenceScreenChanged(title, isInRoot)
    }

    fun goToRootScreen(): Boolean {
        if (isInRoot) return false

        navigateToScreen(null)

        return true
    }

    fun goToSyncScreen() {
        navigateToScreen(findPreference(preferencesHelper.keys.sync) as PreferenceScreen)
    }

    override fun onBackPressed(): Boolean {
        return goToRootScreen()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            onPreferenceScreenChangeListener = context as OnPreferenceScreenChangeListener?
            onPreferenceSyncEnabledChangeListener = context as OnPreferenceSyncEnabledChangeListener?
            onPreferenceClickListener = context as OnPreferenceClickListener?
        } catch (e: ClassCastException) {
            throw ImplementException(context, arrayOf(OnPreferenceScreenChangeListener::class.java, OnPreferenceSyncEnabledChangeListener::class.java, OnPreferenceClickListener::class.java))
        }
    }

    override fun onStart() {
        if (LOG_ENABLED) {
            UtilsLog.d(TAG, "onStart")
        }

        super.onStart()

        init(rootPreferenceScreen)

        rootPreferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onStop() {
        if (LOG_ENABLED) {
            UtilsLog.d(TAG, "onStop")
        }

        rootPreferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)

        super.onStop()
    }

    private fun getValue(value: String?, empty: String): String? {
        return if (TextUtils.isEmpty(value)) empty else value
    }

    private fun updatePreference(key: String) {
        updatePreference(rootPreferenceScreen.findPreference(key))
    }

    @SuppressLint("SwitchIntDef")
    private fun updatePreference(preference: Preference?) {
        if (LOG_ENABLED) {
            UtilsLog.d(TAG, "updatePreference", "preference == $preference")
        }

        val key = preference?.key ?: return

        var title: String? = null
        var summary: String? = null

        when (key) {
            preferencesHelper.keys.price -> summary = getValue(preferencesHelper.priceAsString, getString(R.string.pref_price_empty))
            preferencesHelper.keys.defaultCost,
            preferencesHelper.keys.defaultVolume -> {
                val cost = preferencesHelper.defaultCost
                val volume = preferencesHelper.defaultVolume

                if (cost == 0f && volume == 0f) {
                    summary = if (key == preferencesHelper.keys.defaultCost)
                        getString(R.string.pref_def_cost_empty)
                    else
                        getString(R.string.pref_def_volume_empty)
                } else {
                    val price = preferencesHelper.price

                    if (key == preferencesHelper.keys.defaultCost) {
                        if (cost == 0f) {
                            summary = getString(R.string.pref_def_cost_calc)

                            if (price != 0f) {
                                summary += " (${UtilsFormat.floatToString(volume * price)})"
                            }

                        }
                    } else {
                        if (volume == 0f) {
                            summary = getString(R.string.pref_def_volume_calc)

                            if (price != 0f) {
                                summary += " (${UtilsFormat.floatToString(cost / price)})"
                            }
                        }
                    }
                }

                if (TextUtils.isEmpty(summary)) {
                    summary = preferencesHelper.getString(key)
                }
            }
            preferencesHelper.keys.mapCenterText -> summary = preferencesHelper.mapCenter.text
            preferencesHelper.keys.sync -> summary = getString(if (preferencesHelper.isSyncEnabled)
                R.string.pref_sync_summary_on
            else
                R.string.pref_sync_summary_off)
            preferencesHelper.keys.syncEnabled -> title = getString(if (preferencesHelper.isSyncEnabled)
                R.string.pref_sync_summary_on
            else
                R.string.pref_sync_summary_off)
            preferencesHelper.keys.sms -> summary = getString(if (preferencesHelper.isSMSEnabled)
                R.string.pref_sms_summary_on
            else
                R.string.pref_sms_summary_off)
            preferencesHelper.keys.smsEnabled -> title = getString(if (preferencesHelper.isSMSEnabled)
                R.string.pref_sms_summary_on
            else
                R.string.pref_sms_summary_off)
            preferencesHelper.keys.smsAddress -> summary = preferencesHelper.smsAddress
            preferencesHelper.keys.smsTextPattern -> {
                summary = preferencesHelper.smsTextPattern
                summary = summary.replace("[\\s]+".toRegex(), " ")
            }
            else -> if (preference is EditTextPreference) {
                var text: String? = preferencesHelper.getString(key)

                preference.text = text

                text = getValue(text, "0")

                summary = preference.summary.toString()

                val i = summary.lastIndexOf(" (")

                if (i != -1) {
                    summary = summary.substring(0, i)
                }

                summary += " ($text)"
            }
        }

        if (LOG_ENABLED) {
            UtilsLog.d(TAG, "updatePreference", "title == $title, summary == $summary")
        }

        if (title != null) {
            preference.title = title
        }
        if (summary != null) {
            preference.summary = summary
        }
    }

    private fun init(preference: Preference) {
        updatePreference(preference)
        if (preference is PreferenceScreen || preference is PreferenceGroup) {
            var i = 0
            val count = (preference as PreferenceGroup).preferenceCount
            while (i < count) {
                init(preference.getPreference(i))
                i++
            }
        }
    }

    companion object {
        private const val TAG = "FragmentPreferences"

        private var LOG_ENABLED = false

        const val KEY_PREFERENCE_SCREEN = "KEY_PREFERENCE_SCREEN"
    }
}